source/opt/decoration_manager.cpp

```diff
@@ -231,29 +231,68 @@ std::vector<T> DecorationManager::InternalGetDecorationsFor(
 
 void DecorationManager::ForEachDecoration(uint32_t id,
                                           uint32_t decoration,
-                                          std::function<void(const ir::Instruction&)> f) const {
-  auto decoration_list = id_to_decoration_insts_.find(id);
-  if (decoration_list != id_to_decoration_insts_.end()) {
-    for (const ir::Instruction* inst : decoration_list->second) {
-      switch (inst->opcode()) {
-        case SpvOpDecorate:
-          if (inst->GetSingleWordInOperand(1) == decoration) {
-            f(*inst);
-          }
-          break;
-        case SpvOpMemberDecorate:
-          if (inst->GetSingleWordInOperand(2) == decoration) {
-            f(*inst);
-          }
-          break;
-        case SpvOpDecorateId:
-          if (inst->GetSingleWordInOperand(1) == decoration) {
-            f(*inst);
+                                          std::function<void(const ir::Instruction&)> f) {
+  for (const ir::Instruction* inst : GetDecorationsFor(id, true)) {
+    switch (inst->opcode()) {
+      case SpvOpMemberDecorate:
+        if (inst->GetSingleWordInOperand(2) == decoration) {
+          f(*inst);
+        }
+        break;
+      case SpvOpDecorate:
+      case SpvOpDecorateId:
+        if (inst->GetSingleWordInOperand(1) == decoration) {
+          f(*inst);
+        }
+        break;
+      default:
+        assert(false && "Unexpected decoration instruction");
+    }
+  }
+}
+
+void DecorationManager::CloneDecorations(uint32_t from, uint32_t to,
+                                         std::function<void(ir::Instruction&, bool)> f) {
+  auto const decoration_list = id_to_decoration_insts_.find(from);
+  if (decoration_list == id_to_decoration_insts_.end()) return;
+  for (ir::Instruction* inst : decoration_list->second) {
+    switch (inst->opcode()) {
+      case SpvOpGroupDecorate:
+        f(*inst, false);
+        // add |to| to list of decorated id's
+        inst->AddOperand(std::move(
+          ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID, {to})));
+        id_to_decoration_insts_[to].push_back(inst);
+        f(*inst, true);
+        break;
+      case SpvOpGroupMemberDecorate:
+        f(*inst, false);
+        // for each (id == from), add (to, literal) as operands
+        for (uint32_t i = 1; i < inst->NumOperands(); i+=2) {
```

Comment:
Consider
const uint32_t num_operand =  inst->NumOperands();
for (uint32_t i = 1; i < num_operands; i+=2) {
since you don't need to iterate over operands you are adding.